File: Components/DevExpressVCL/Demos/ExpressQuantumGrid/Delphi/EditorsInPlaceDemo/EditorsInPlaceDemoMain.pas
unit EditorsInPlaceDemoMain;

interface

uses
  Windows, Messages, Classes, Graphics, Controls, Forms,
  Dialogs, StdCtrls, cxGridLevel, cxGridCustomTableView, cxGridTableView,
  cxGridDBTableView, cxControls, cxGridCustomView, cxGrid, cxCustomData,
  ExtCtrls, ActnList, ImgList, Menus, ComCtrls, cxGridCustomPopupMenu,
  cxGridPopupMenu, Grids, DBGrids, cxDBData, cxStyles, cxGraphics, cxFilter,
  cxData, cxEdit, DB, cxClasses, cxGridCardView, cxGridDBCardView, cxDBEdit, cxDropDownEdit,
  cxDataStorage, cxMRUEdit, cxButtonEdit, cxCheckBox, cxMaskEdit, cxMemo,
  cxDBLookupComboBox, cxTimeEdit, cxImageComboBox, cxSpinEdit, cxCalc,
  cxImage, cxHyperLinkEdit, cxLookAndFeels, cxLookAndFeelPainters,
  cxGridCustomLayoutView, BaseForm, CarsDataForGrid, cxNavigator, dxBarCode;

type
  TEditorsInPlaceDemoMainForm = class(TfmBaseForm)
    miOptions: TMenuItem;
    Grid: TcxGrid;
    lvCustomers: TcxGridLevel;
    miShowEditButtons: TMenuItem;
    miEditButtonsNever: TMenuItem;
    miEditButtonsForFocusedRecord: TMenuItem;
    miEditButtonsAlways: TMenuItem;
    lvOrders: TcxGridLevel;
    lvCars: TcxGridLevel;
    cvCustomers: TcxGridDBCardView;
    cvOrders: TcxGridDBCardView;
    cvCars: TcxGridDBCardView;
    cvCustomersFirstName: TcxGridDBCardViewRow;
    cvCustomersLastName: TcxGridDBCardViewRow;
    cvCustomersCompany: TcxGridDBCardViewRow;
    cvCustomersPrefix: TcxGridDBCardViewRow;
    cvCustomersAddress: TcxGridDBCardViewRow;
    cvCustomersCity: TcxGridDBCardViewRow;
    cvCustomersState: TcxGridDBCardViewRow;
    cvCustomersZipCode: TcxGridDBCardViewRow;
    cvCustomersCustomer: TcxGridDBCardViewRow;
    cvCustomersFaxPhone: TcxGridDBCardViewRow;
    cvCustomersDescription: TcxGridDBCardViewRow;
    cvOrdersCustomerID: TcxGridDBCardViewRow;
    cvOrdersProductID: TcxGridDBCardViewRow;
    cvOrdersPurchaseDate: TcxGridDBCardViewRow;
    cvOrdersTime: TcxGridDBCardViewRow;
    cvOrdersPaymentType: TcxGridDBCardViewRow;
    cvOrdersDescription: TcxGridDBCardViewRow;
    cvOrdersQuantity: TcxGridDBCardViewRow;
    cvOrdersPaymentAmount: TcxGridDBCardViewRow;
    cvCarsPicture: TcxGridDBCardViewRow;
    cvCarsCar: TcxGridDBCardViewRow;
    cvCarsCategory: TcxGridDBCardViewRow;
    cvCarsHyperlink: TcxGridDBCardViewRow;
    cvCarsInfo: TcxGridDBCardViewRow;
    cvBarCode: TcxGridDBCardViewRow;
    procedure FormShow(Sender: TObject);
    procedure cvCustomersStatePropertiesButtonClick(Sender: TObject);
    procedure cvCustomersCityPropertiesButtonClick(Sender: TObject;
      AButtonIndex: Integer);
    procedure GridFocusedViewChanged(Sender: TcxCustomGrid;
      APrevFocusedView, AFocusedView: TcxCustomGridView);
    procedure miEditButtonsClick(Sender: TObject);
    procedure cvCarsCustomDrawCell(Sender: TcxCustomGridTableView;
      ACanvas: TcxCanvas; AViewInfo: TcxGridTableDataCellViewInfo;
      var ADone: Boolean);
  private
    procedure UpdateOptions;
  protected
    procedure LookAndFeelChanged(Sender: TcxLookAndFeel;
      AChangedValues: TcxLookAndFeelValues); override;
  end;

var
  EditorsInPlaceDemoMainForm: TEditorsInPlaceDemoMainForm;

implementation

{$R *.dfm}

uses
  EditorsInPlaceDemoData, ShellAPI,
  EditorsInPlaceDemoCities, EditorsInPlaceDemoCars, AboutDemoForm;

procedure TEditorsInPlaceDemoMainForm.FormShow(Sender: TObject);
begin
  TcxPopupEditProperties(cvCarsInfo.Properties).ImmediatePopup := True;
end;

procedure TEditorsInPlaceDemoMainForm.cvCustomersStatePropertiesButtonClick(
  Sender: TObject);
begin
  MessageDlg('Button click event handler', mtInformation, [mbOK], 0);
end;

procedure TEditorsInPlaceDemoMainForm.cvCustomersCityPropertiesButtonClick(
  Sender: TObject; AButtonIndex: Integer);
begin
  case AButtonIndex of
    0:
      if EditorsInPlaceDemoCitiesForm.ShowModal = mrOk then
      begin
        EditorsInPlaceDemoDataDM.tblCustomers.Edit;
        EditorsInPlaceDemoDataDM.tblCustomersCity.Value := EditorsInPlaceDemoCitiesForm.Value;
      end;
    1: ShellExecute(Handle, 'OPEN', 'http://www.usacitiesonline.com/', nil, nil, SW_SHOWMAXIMIZED);
  end;
end;

procedure TEditorsInPlaceDemoMainForm.GridFocusedViewChanged(
  Sender: TcxCustomGrid; APrevFocusedView, AFocusedView: TcxCustomGridView);
begin
  UpdateOptions;
end;

procedure TEditorsInPlaceDemoMainForm.UpdateOptions;
var
  AView: TcxGridDBCardView;
begin
  AView := TcxGridDBCardView(Grid.FocusedView);
  case AView.OptionsView.ShowEditButtons of
    gsebNever:
      MenuItemSetChecked('miEditButtonsNever', True);
    gsebForFocusedRecord:
      MenuItemSetChecked('miEditButtonsForFocusedRecord', True);
    gsebAlways:
      MenuItemSetChecked('miEditButtonsAlways', True);
  end;
end;

procedure TEditorsInPlaceDemoMainForm.LookAndFeelChanged(
  Sender: TcxLookAndFeel; AChangedValues: TcxLookAndFeelValues);
begin
  inherited LookAndFeelChanged(Sender, AChangedValues);
  UpdateCardViewStyleSheet(cvCustomers);
  UpdateCardViewStyleSheet(cvOrders);
  UpdateCardViewStyleSheet(cvCars);
end;

procedure TEditorsInPlaceDemoMainForm.miEditButtonsClick(Sender: TObject);
var
  AView: TcxGridDBCardView;
begin
  AView := TcxGridDBCardView(Grid.FocusedView);
  case (Sender as TComponent).Tag of
    0: AView.OptionsView.ShowEditButtons := gsebNever;
    1: AView.OptionsView.ShowEditButtons := gsebForFocusedRecord;
    2: AView.OptionsView.ShowEditButtons := gsebAlways;
  end;
  MenuItemSetChecked(Sender, True);
end;

procedure TEditorsInPlaceDemoMainForm.cvCarsCustomDrawCell(
  Sender: TcxCustomGridTableView; ACanvas: TcxCanvas;
  AViewInfo: TcxGridTableDataCellViewInfo; var ADone: Boolean);
begin
  if (AViewInfo.Item.Index = cvCarsInfo.Index) and (AViewInfo is TcxGridCardRowDataViewInfo) then
  begin
    ACanvas.Canvas.Brush.Style := bsSolid;
    ACanvas.Canvas.Brush.Color := $00F7EAD9;
    ACanvas.Canvas.FillRect(AViewInfo.Bounds);
    SetBkMode(ACanvas.Handle, TRANSPARENT);
    ACanvas.Font.Color := clBlack;
    ACanvas.DrawText('<click here and view dropdown window>', AViewInfo.Bounds, 0);
    ADone := True;
  end;
end;

end.

File: Components/DevExpressVCL/Demos/ExpressQuantumGrid/CBuilder/EditorsInPlaceDemo/EditorsInPlaceDemoMain.cpp
//---------------------------------------------------------------------------

#include <vcl.h>
#include "shellapi.hpp"
#pragma hdrstop

#include "EditorsInPlaceDemoMain.h"
#include "EditorsInPlaceDemoData.h"
#include "AboutDemoForm.h"
#include "EditorsInPlaceDemoCities.h"
#include "EditorsInPlaceDemoCars.h"

//---------------------------------------------------------------------------
#pragma package(smart_init)
#pragma link "cxClasses"
#pragma link "cxControls"
#pragma link "cxCustomData"
#pragma link "cxData"
#pragma link "cxDBData"
#pragma link "cxEdit"
#pragma link "cxFilter"
#pragma link "cxGraphics"
#pragma link "cxGrid"
#pragma link "cxGridCustomTableView"
#pragma link "cxGridCustomView"
#pragma link "cxGridDBTableView"
#pragma link "cxGridLevel"
#pragma link "cxGridTableView"
#pragma link "cxStyles"
#pragma link "cxGridCardView"
#pragma link "cxGridDBCardView"
#pragma link "cxLookAndFeels"
#pragma link "cxButtonEdit"
#pragma link "cxCalc"
#pragma link "cxCheckBox"
#pragma link "cxDataStorage"
#pragma link "cxDBLookupComboBox"
#pragma link "cxDropDownEdit"
#pragma link "cxGridCustomLayoutView"
#pragma link "cxHyperLinkEdit"
#pragma link "cxImage"
#pragma link "cxImageComboBox"
#pragma link "cxLookAndFeelPainters"
#pragma link "cxMaskEdit"
#pragma link "cxMemo"
#pragma link "cxMRUEdit"
#pragma link "cxSpinEdit"
#pragma link "cxTimeEdit"
#pragma resource "*.dfm"
TEditorsInPlaceDemoMainForm *EditorsInPlaceDemoMainForm;
//---------------------------------------------------------------------------
__fastcall TEditorsInPlaceDemoMainForm::TEditorsInPlaceDemoMainForm(TComponent* Owner)
  : TfmBaseForm(Owner)
{
}
//---------------------------------------------------------------------------

void __fastcall TEditorsInPlaceDemoMainForm::FormShow(TObject *Sender)
{
  ((TcxPopupEditPropertiesAccess*)cvCarsInfo->Properties)->ImmediatePopup = true;
}
//---------------------------------------------------------------------------

void __fastcall TEditorsInPlaceDemoMainForm::cvCustomersStatePropertiesButtonClick(TObject *Sender)
{
  MessageDlg("Button click event handler", mtInformation, TMsgDlgButtons()<<mbOK, 0);
}
//---------------------------------------------------------------------------

void __fastcall TEditorsInPlaceDemoMainForm::cvCustomersCityPropertiesButtonClick(TObject *Sender,
    int AButtonIndex)
{
  switch (AButtonIndex){
    case 0:
      if (EditorsInPlaceDemoCitiesForm->ShowModal() == mrOk){
        EditorsInPlaceDemoDataDM->tblCustomers->Edit();
        EditorsInPlaceDemoDataDM->tblCustomersCity->Value = EditorsInPlaceDemoCitiesForm->Value;
      }
      break;
    case 1:
      ShellExecute(Handle, PAnsiChar("OPEN"), PAnsiChar("http://www.usacitiesonline.com/"), NULL, NULL, SW_SHOWMAXIMIZED);
  }
}
//---------------------------------------------------------------------------

void __fastcall TEditorsInPlaceDemoMainForm::GridFocusedViewChanged(TcxCustomGrid *Sender,
    TcxCustomGridView *APrevFocusedView, TcxCustomGridView *AFocusedView)
{
  UpdateOptions();
}
//---------------------------------------------------------------------------

void __fastcall TEditorsInPlaceDemoMainForm::miEditButtonsClick(TObject *Sender)
{
  if (!((TMenuItem*)Sender)->Checked) {
    TcxGridDBCardView *AView = (TcxGridDBCardView*)Grid->FocusedView;
    switch (((TMenuItem*)Sender)->Tag){
      case 0:
        AView->OptionsView->ShowEditButtons = gsebNever;
        break;
      case 1:
        AView->OptionsView->ShowEditButtons = gsebForFocusedRecord;
        break;
      case 2:
        AView->OptionsView->ShowEditButtons = gsebAlways;
    }
    ((TMenuItem*)Sender)->Checked = true;
  }
}
//---------------------------------------------------------------------------

void __fastcall TEditorsInPlaceDemoMainForm::cvCarsCustomDrawCell(TcxCustomGridTableView *Sender,
    TcxCanvas *ACanvas, TcxGridTableDataCellViewInfo *AViewInfo,
    bool &ADone)
{
  if ((AViewInfo->Item->Index == cvCarsInfo->Index) &&
    (dynamic_cast< TcxGridCardRowDataViewInfo* >(AViewInfo) != 0)){
    ACanvas->Canvas->Brush->Style = bsSolid;
    ACanvas->Canvas->Brush->Color = (TColor)0x00F7EAD9;
    ACanvas->Canvas->FillRect(AViewInfo->Bounds);
    SetBkMode(ACanvas->Handle, TRANSPARENT);
    ACanvas->Font->Color = clBlack;
    ACanvas->Canvas->TextOut(AViewInfo->Bounds.Left, AViewInfo->Bounds.Top, "<click here and view dropdown window>");
    ADone = true;
  }
}
//---------------------------------------------------------------------------

void TEditorsInPlaceDemoMainForm::UpdateOptions()
{
  switch (((TcxGridDBCardView*)Grid->FocusedView)->OptionsView->ShowEditButtons){
  case gsebNever:
    miEditButtonsNever->Checked = true;
    break;
  case gsebForFocusedRecord:
    miEditButtonsForFocusedRecord->Checked = true;
    break;
  case gsebAlways:
    miEditButtonsAlways->Checked = true;
 }
}
//---------------------------------------------------------------------------

